crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+        let lookup_error = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => return Self::read_zipfile(zipfile),
+            Err(err) => err,
+        };
+        if normalized.has_trailing_slash() {
+            return Err(lookup_error);
+        }
+        let mut archive = self.vendored_archive();
+        let zipfile = archive.lookup_path(&normalized.with_trailing_slash())?;
+        Self::read_zipfile(zipfile)
+    }
+}
+
+/// Newtype wrapper around a ZipArchive.
+///
+/// Since ZipArchives are generally cheap to clone (for now), this should be too.
+#[derive(Debug, Clone)]
+struct VendoredZipArchive(ZipArchive<Cursor<Arc<[u8]>>>);
+
+impl VendoredZipArchive {
+    fn new(vendored_bytes: Arc<[u8]>) -> ZipResult<Self> {
+        Ok(Self(ZipArchive::new(Cursor::new(vendored_bytes))?))
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
+}
+
+/// A path that has been normalized via the `normalize_vendored_path` function.
+///
+/// Trailing slashes are normalized away by `camino::Utf8PathBuf`s,
+/// but trailing slashes are crucial for distinguishing between
+/// files and directories inside zip archives.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NormalizedVendoredPath(String);
+
+impl NormalizedVendoredPath {
+    fn with_trailing_slash(mut self) -> Self {
+        self.0.push('/');
+        self
+    }
+
+    fn has_trailing_slash(&self) -> bool {
+        self.0.ends_with('/')
+    }
+
+    fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Normalizes the path by removing `.` and `..` components.
+///
+/// ## Panics:
+/// If a path with an unsupported component for vendored paths is passed.
+/// Unsupported components are path prefixes,
+/// and path root directories appearing anywhere except at the start of the path.
+fn normalize_vendored_path(path: &FileSystemPath) -> NormalizedVendoredPath {
+    debug_assert!(
+        !path.as_std_path().is_absolute(),
+        "Attempted to lookup an absolute filesystem path relative to a vendored zip archive"
+    );
+    let mut normalized_parts = camino::Utf8PathBuf::new();
+
+    // Allow the `RootDir` component, but only if it is at the very start of the string.
+    let mut components = path.components().peekable();
+    if let Some(camino::Utf8Component::RootDir) = components.peek() {
+        components.next();
+    }
+
+    for component in components {
+        match component {
+            camino::Utf8Component::Normal(part) => normalized_parts.push(part),
+            camino::Utf8Component::CurDir => continue,
+            camino::Utf8Component::ParentDir => {
+                normalized_parts.pop();
+            }
+            unsupported => panic!("Unsupported component in a vendored path: {unsupported}"),
+        }
+    }
```

Comment:
We have almost the same logic in `memory` rs 😆

Reply:
Not a coincidence 😉 I looked at that function quite a bit while working on this, but in the end I concluded that there was enough that needed to be different that it would require a separate function.